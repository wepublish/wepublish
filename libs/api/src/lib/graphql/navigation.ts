import {
  GraphQLInterfaceType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
  GraphQLUnionType,
  GraphQLID,
  GraphQLInputObjectType
} from 'graphql'

import {Context} from '../context'

import {
  ArticleNavigationLink,
  NavigationLinkType,
  PageNavigationLink,
  ExternalNavigationLink
} from '../db/navigation'
import {createProxyingResolver, createProxyingIsTypeOf} from '../utility'
import {Page} from '@wepublish/page/api'
import {Article} from '@wepublish/article/api'

export const GraphQLBaseNavigationLink = new GraphQLInterfaceType({
  name: 'BaseNavigationLink',
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)}
  }
})

export const GraphQLPageNavigationLink = new GraphQLObjectType<PageNavigationLink, Context>({
  name: 'PageNavigationLink',
  interfaces: [GraphQLBaseNavigationLink],
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    page: {
      type: Page,
      resolve: createProxyingResolver(({pageID}, _args, {loaders}) => {
        return {
          __typename: 'Page',
          id: pageID
        }
      })
    }
  },
  isTypeOf: createProxyingIsTypeOf(value => {
    return value.type === NavigationLinkType.Page
  })
})

export const GraphQLArticleNavigationLink = new GraphQLObjectType<ArticleNavigationLink, Context>({
  name: 'ArticleNavigationLink',
  interfaces: [GraphQLBaseNavigationLink],
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    article: {
      type: Article,
      resolve: createProxyingResolver(({articleID}, _args, {loaders}) => {
        return {
          __typename: 'Article',
          id: articleID
        }
      })
    }
  },
  isTypeOf: createProxyingIsTypeOf(value => {
    return value.type === NavigationLinkType.Article
  })
})

export const GraphQLExternalNavigationLink = new GraphQLObjectType<ExternalNavigationLink, Context>(
  {
    name: 'ExternalNavigationLink',
    interfaces: [GraphQLBaseNavigationLink],
    fields: {
      label: {type: new GraphQLNonNull(GraphQLString)},
      url: {type: new GraphQLNonNull(GraphQLString)}
    },
    isTypeOf: createProxyingIsTypeOf(value => {
      return value.type === NavigationLinkType.External
    })
  }
)

export const GraphQLNavigationLink = new GraphQLUnionType({
  name: 'NavigationLink',
  types: [GraphQLPageNavigationLink, GraphQLArticleNavigationLink, GraphQLExternalNavigationLink]
})

export const GraphQLNavigation = new GraphQLObjectType({
  name: 'Navigation',
  fields: {
    id: {type: new GraphQLNonNull(GraphQLID)},
    key: {type: new GraphQLNonNull(GraphQLString)},
    name: {type: new GraphQLNonNull(GraphQLString)},
    links: {type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLNavigationLink)))}
  }
})

export const GraphQLPublicPageNavigationLink = new GraphQLObjectType<PageNavigationLink, Context>({
  name: 'PageNavigationLink',
  interfaces: [GraphQLBaseNavigationLink],
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    page: {
      type: Page,
      resolve: createProxyingResolver(({pageID}, _args, {loaders}) => {
        return {
          __typename: 'Page',
          id: pageID
        }
      })
    }
  },
  isTypeOf: createProxyingIsTypeOf(value => {
    return value.type === NavigationLinkType.Page
  })
})

export const GraphQLPublicArticleNavigationLink = new GraphQLObjectType<
  ArticleNavigationLink,
  Context
>({
  name: 'ArticleNavigationLink',
  interfaces: [GraphQLBaseNavigationLink],
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    article: {
      type: Article,
      resolve: createProxyingResolver(({articleID}, _args, {loaders}) => {
        return {
          __typename: 'Article',
          id: articleID
        }
      })
    }
  },
  isTypeOf: createProxyingIsTypeOf(value => {
    return value.type === NavigationLinkType.Article
  })
})

export const GraphQLPublicNavigationLink = new GraphQLUnionType({
  name: 'NavigationLink',
  types: [
    GraphQLPublicPageNavigationLink,
    GraphQLPublicArticleNavigationLink,
    GraphQLExternalNavigationLink
  ]
})

export const GraphQLPublicNavigation = new GraphQLObjectType({
  name: 'Navigation',
  fields: {
    id: {type: new GraphQLNonNull(GraphQLID)},
    key: {type: new GraphQLNonNull(GraphQLString)},
    name: {type: new GraphQLNonNull(GraphQLString)},
    links: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLPublicNavigationLink)))
    }
  }
})

export const GraphQLArticleNavigationLinkInput = new GraphQLInputObjectType({
  name: 'ArticleNavigationLinkInput',
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    articleID: {type: new GraphQLNonNull(GraphQLID)}
  }
})

export const GraphQLPageNavigationLinkInput = new GraphQLInputObjectType({
  name: 'PageNavigationLinkInput',
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    pageID: {type: new GraphQLNonNull(GraphQLID)}
  }
})

export const GraphQLExternalNavigationLinkInput = new GraphQLInputObjectType({
  name: 'ExternalNavigationLinkInput',
  fields: {
    label: {type: new GraphQLNonNull(GraphQLString)},
    url: {type: new GraphQLNonNull(GraphQLString)}
  }
})

export const GraphQLNavigationLinkInput = new GraphQLInputObjectType({
  name: 'NavigationLinkInput',
  fields: {
    [NavigationLinkType.Page]: {type: GraphQLPageNavigationLinkInput},
    [NavigationLinkType.Article]: {type: GraphQLArticleNavigationLinkInput},
    [NavigationLinkType.External]: {type: GraphQLExternalNavigationLinkInput}
  }
})

export const GraphQLNavigationInput = new GraphQLInputObjectType({
  name: 'NavigationInput',
  fields: {
    key: {type: new GraphQLNonNull(GraphQLString)},
    name: {type: new GraphQLNonNull(GraphQLString)},
    links: {
      type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(GraphQLNavigationLinkInput)))
    }
  }
})
