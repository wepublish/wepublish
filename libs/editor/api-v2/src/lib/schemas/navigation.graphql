fragment FullNavigation on Navigation {
    id
    key
    name
    links {
        __typename

        ... on PageNavigationLink {
            label
            pageID
        }
        ... on ArticleNavigationLink {
            label
            articleID
        }
        ... on ExternalNavigationLink {
            label
            url
        }
    }
}

query NavigationList {
    getNavigations {
        ...FullNavigation
    }
}

query Navigation($id: ID!) {
    getNavigationById(id: $id) {
        ...FullNavigation
    }
}

mutation CreateNavigation($navigation: CreateNavigationInput!) {
    createNavigation(navigation: $navigation) {
        ...FullNavigation
    }
}

mutation UpdateNavigation($navigation: UpdateNavigationInput!) {
    updateNavigation(navigation: $navigation) {
        ...FullNavigation
    }
}

mutation DeleteNavigation($id: ID!) {
    deleteNavigation(id: $id) {
        ...FullNavigation
    }
}
