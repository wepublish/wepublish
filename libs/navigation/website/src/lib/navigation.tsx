import styled from '@emotion/styled'
import AppBar from '@mui/material/AppBar'
import Box from '@mui/material/Box'
import Toolbar from '@mui/material/Toolbar'
import Typography from '@mui/material/Typography'
import {BuilderNavigationProps, useWebsiteBuilder} from '@wepublish/website-builder'

const NavigationWrapper = styled(Box)`
  flex-grow: 1;
`

const Title = styled.div`
  flex-grow: 1;
`

export function Navigation({data, loading, error}: BuilderNavigationProps) {
  const {Button} = useWebsiteBuilder()

  return (
    <NavigationWrapper>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="h6" component={Title}>
            Newsroom
          </Typography>

          <Button>Login</Button>
        </Toolbar>
      </AppBar>
    </NavigationWrapper>
  )
}
