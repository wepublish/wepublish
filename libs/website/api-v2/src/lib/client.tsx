import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
