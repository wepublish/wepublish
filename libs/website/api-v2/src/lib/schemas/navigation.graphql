fragment FullNavigation on Navigation {
    id
    key
    name
    links {
        ... on PageNavigationLink {
            label
            pageID
        }

        ... on ArticleNavigationLink {
            label
            articleID
        }

        ... on ExternalNavigationLink {
            label
            url
        }
    }
}

query NavigationList {
    getNavigations {
        ...FullNavigation
    }
}

query Navigation($slug: String!) {
    getNavigationByKey(key: $slug) {
        ...FullNavigation
    }
}
