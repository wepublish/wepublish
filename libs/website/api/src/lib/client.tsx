import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
