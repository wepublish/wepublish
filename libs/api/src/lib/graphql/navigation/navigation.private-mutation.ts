import {Context} from '../../context'
import {authorise} from '../permissions'
import {CanCreateNavigation, CanDeleteNavigation} from '@wepublish/permissions/api'
import {PrismaClient, Prisma} from '@prisma/client'

export const deleteNavigationById = (
  id: string,
  authenticate: Context['authenticate'],
  navigation: PrismaClient['navigation']
) => {
  const {roles} = authenticate()
  authorise(CanDeleteNavigation, roles)

  return navigation.delete({
    where: {
      id
    },
    include: {
      links: true
    }
  })
}

type CreateNavigationInput = Omit<Prisma.NavigationUncheckedCreateInput, 'links' | 'modifiedAt'> & {
  links: Prisma.NavigationLinkUncheckedCreateWithoutNavigationInput[]
}

export const createNavigation = (
  {links, ...input}: CreateNavigationInput,
  authenticate: Context['authenticate'],
  navigation: PrismaClient['navigation']
) => {
  const {roles} = authenticate()
  authorise(CanCreateNavigation, roles)

  return navigation.create({
    data: {
      ...input,
      links: {
        createMany: {
          data: links
        }
      }
    },
    include: {
      links: true
    }
  })
}

type UpdateNavigationInput = Omit<
  Prisma.NavigationUncheckedUpdateInput,
  'links' | 'modifiedAt' | 'createdAt'
> & {
  links: Prisma.NavigationLinkUncheckedCreateWithoutNavigationInput[]
}

export const updateNavigation = async (
  id: string,
  {links, ...input}: UpdateNavigationInput,
  authenticate: Context['authenticate'],
  navigation: PrismaClient['navigation']
) => {
  const {roles} = authenticate()
  authorise(CanCreateNavigation, roles)

  return navigation.update({
    where: {id},
    data: {
      ...input,
      links: {
        deleteMany: {
          navigationId: {
            equals: id
          }
        },
        createMany: {
          data: links
        }
      }
    },
    include: {
      links: true
    }
  })
}
