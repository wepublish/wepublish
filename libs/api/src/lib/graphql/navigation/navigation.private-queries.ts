import {Context} from '../../context'
import {UserInputError} from '../../error'
import {authorise} from '../permissions'
import {CanGetNavigation, CanGetNavigations} from '@wepublish/permissions/api'
import {PrismaClient} from '@prisma/client'

export const getNavigationByIdOrKey = (
  id: string | null,
  key: string | null,
  authenticate: Context['authenticate'],
  navigationByID: Context['loaders']['navigationByID'],
  navigationByKey: Context['loaders']['navigationByKey']
) => {
  const {roles} = authenticate()
  authorise(CanGetNavigation, roles)

  if ((!id && !key) || (id && key)) {
    throw new UserInputError('You must provide either `id` or `key`.')
  }

  return id ? navigationByID.load(id) : navigationByKey.load(key!)
}

export const getNavigations = (
  authenticate: Context['authenticate'],
  navigation: PrismaClient['navigation']
) => {
  const {roles} = authenticate()
  authorise(CanGetNavigations, roles)

  return navigation.findMany({
    orderBy: {
      createdAt: 'desc'
    },
    include: {
      links: true
    }
  })
}
