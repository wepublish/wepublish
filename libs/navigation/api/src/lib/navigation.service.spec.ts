import {Test, TestingModule} from '@nestjs/testing'
import {PrismaClient} from '@prisma/client'
import {NavigationService} from './navigation.service'
import {NavigationDataloader} from './navigation.dataloader'

describe('NavigationService', () => {
  let service: NavigationService
  let navigationDataloader: {[method in keyof NavigationDataloader]?: jest.Mock}
  let prismaMock: any

  beforeEach(async () => {
    navigationDataloader = {
      load: jest.fn()
    }

    prismaMock = {
      navigation: {
        findUnique: jest.fn(),
        findMany: jest.fn(),
        create: jest.fn(),
        delete: jest.fn(),
        update: jest.fn()
      },
      navigationLink: {
        deleteMany: jest.fn()
      }
    }

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        NavigationService,
        {provide: NavigationDataloader, useValue: navigationDataloader},
        {provide: PrismaClient, useValue: prismaMock}
      ]
    }).compile()

    service = module.get<NavigationService>(NavigationService)
  })

  it('should get a navigation by id', async () => {
    prismaMock.navigation.findUnique.mockResolvedValueOnce({})
    await service.getNavigationById('123')
    expect(prismaMock.navigation.findUnique).toHaveBeenCalled()
    expect(prismaMock.navigation.findUnique.mock.calls[0]).toMatchSnapshot()
  })

  it('should get a navigation by key', async () => {
    prismaMock.navigation.findUnique.mockResolvedValueOnce({})
    await service.getNavigationByKey('key123')
    expect(prismaMock.navigation.findUnique).toHaveBeenCalled()
    expect(prismaMock.navigation.findUnique.mock.calls[0]).toMatchSnapshot()
  })

  it('should get all navigations', async () => {
    prismaMock.navigation.findMany.mockResolvedValueOnce([])
    await service.getNavigations()
    expect(prismaMock.navigation.findMany).toHaveBeenCalled()
    expect(prismaMock.navigation.findMany.mock.calls[0]).toMatchSnapshot()
  })

  it('should create a navigation', async () => {
    const input = {key: 'testKey', name: 'testName', links: []}
    prismaMock.navigation.create.mockResolvedValueOnce({})
    await service.createNavigation(input)
    expect(prismaMock.navigation.create).toHaveBeenCalled()
    expect(prismaMock.navigation.create.mock.calls[0]).toMatchSnapshot()
  })

  it('should delete a navigation by id', async () => {
    prismaMock.navigation.delete.mockResolvedValueOnce({})
    await service.deleteNavigationById('123')
    expect(prismaMock.navigation.delete).toHaveBeenCalled()
    expect(prismaMock.navigation.delete.mock.calls[0]).toMatchSnapshot()
  })

  it('should update a navigation', async () => {
    const input = {id: '123', key: 'testKey', name: 'testName', links: []}
    prismaMock.navigation.update.mockResolvedValueOnce({})
    prismaMock.navigationLink.deleteMany.mockResolvedValueOnce({})
    await service.updateNavigation(input)
    expect(prismaMock.navigationLink.deleteMany).toHaveBeenCalled()
    expect(prismaMock.navigation.update).toHaveBeenCalled()
    expect(prismaMock.navigation.update.mock.calls[0]).toMatchSnapshot()
  })
})
