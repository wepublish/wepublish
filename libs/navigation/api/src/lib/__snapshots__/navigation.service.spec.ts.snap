// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationService should create a navigation 1`] = `
[
  {
    "data": {
      "key": "testKey",
      "links": {
        "createMany": {
          "data": [],
        },
      },
      "name": "testName",
    },
    "include": {
      "links": true,
    },
  },
]
`;

exports[`NavigationService should delete a navigation by id 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "where": {
      "id": "123",
    },
  },
]
`;

exports[`NavigationService should get a navigation by id 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "where": {
      "id": "123",
    },
  },
]
`;

exports[`NavigationService should get a navigation by key 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "where": {
      "key": "key123",
    },
  },
]
`;

exports[`NavigationService should get all navigations 1`] = `
[
  {
    "include": {
      "links": true,
    },
    "orderBy": {
      "createdAt": "desc",
    },
  },
]
`;

exports[`NavigationService should update a navigation 1`] = `
[
  {
    "data": {
      "key": "testKey",
      "links": {
        "createMany": {
          "data": [],
        },
      },
      "name": "testName",
    },
    "include": {
      "links": true,
    },
    "where": {
      "id": "123",
    },
  },
]
`;
